### ISSUE TITLE
Errors in global registration are likely to be eaten

### ISSUE THREAD

**Issue by Rauno56** [open-telemetry/opentelemetry-js-api#56]
https://github.com/open-telemetry/opentelemetry-js-api/blob/main/src/internal/global-utils.ts uses diag logger to report issues with registering globals.

This is not adequate when the user doesn't control both of the installed packages registering those globals(that is definitely most cases) because they would then have no way to configure the logger to actually show those messages in the first place unless the logger they have access to is used for the second attempt to register the global.

**Comment by dyladan** [comment]
What would you suggest?

**Comment by Rauno56** [comment]
I'm getting myself a little more familiar with what's there and understand the whys better before I come back with a suggestion.

**Comment by dyladan** [comment]
There has been a previous suggestion for a debug env var which would automatically enable the diag logger to log to console

**Comment by Rauno56** [comment]
The first thing that came to my mind was to just log the error to **every** registered logger.

For loggers themselves, since there's ever only one logger for one version it should also be done just before any logger is overwritten so if there's another SDK on the same version that overwrites my logger I should be notified as well as the new logger that such an event occurred.

That's a quick fix, but I do not think it's the whole story... :thinking:

**Comment by dyladan** [comment]
Can you provide an example usecase where the user doesn't have control over registering the globals? Only the end-user should ever do this so I'm not sure how realistic the situation is.

**Comment by Rauno56** [comment]
I deal with the issue every day writing and debugging instrumentations. Since one of the instrumentations is linked, it has a separate dep tree on filesystem. Can even be the same version API, but different instances of it - There's one installed in the developed instrumentation `node_modules` and another in my example code using the developed instrumentation.

The reason current setup works most of the time, is because npm dedupes the API package. I'm sure there are other reasons why that might not happen, but relying on that feature is potentially dangerous imo.

1. I don't know a good way to fix it for myself right now;
2. Theres currently no visibility for me to see what and when actually happens.

**Comment by Rauno56** [comment]
On the other side, I'm trying to think of an usecase where having multiple APIs registered would have unexpected consequences:

Say you install and SDK and few instrumentations which require different versions of the API. Currently some of those packages would get an access to context manager I, as a user configured, others would likely not, because my SDK uses and configures only one of those sets of global pointers.

@dyladan, could you please explain a usecase where having parallel APIs would be useful?

**Comment by dyladan** [comment]
>I deal with the issue every day writing and debugging instrumentations. Since one of the instrumentations is linked, it has a separate dep tree on filesystem. Can even be the same version API, but different instances of it - There's one installed in the developed instrumentation node_modules and another in my example code using the developed instrumentation.

We recommend peer depending on the API for this reason

**Comment by vmarchaud** [comment]
Yes they should surface with npm but some people can still use older npm version so thats not a catch all solution, just hoping it would in the future :)

**Comment by dyladan** [comment]
It also doesn't work for third parties that want to build in official support. If the mysql driver for example built in support for otel, they wouldn't want to peer depend on the API and cause warnings for their customers that don't use otel.

**Comment by glasser** [comment]
> It is not useful. We recommend peer depending on the API for this reason

@dyladan Are these recommendations written somewhere? We're trying to figure out the best way to add some spans to a library of ours and aren't entirely sure if we should be calling `getTracer` in the library vs having users inject it, doing direct vs peer dependencies on `@opentelemetry/api`, etc. We know about the built-in version checks but (especially in this pre-1.0 world where there's a semver-incompatible release every few weeks) that feels like it is just going to lead to surprising no-ops unless we are careful?  Or is it possible for a single library to work with multiple 0-minor versions being careful about the API changes with conditionals? Would love to be able to read a recommended guide to how to integrate tracer into your library.

**Comment by vmarchaud** [comment]
@glasser We released the 1.0 which lock the trace/context API (even though you can't use the SDK with it right now, should be in the coming week). Otherwise do you have more information about the library ? Is it a opentelemetry focused library that you can re-use across your service (in which case i would advise to directly depend on the API) or a library that anyone can use (in which case as dyladan said, it should directly on the api to avoid warning for people that doesn't use otel)

**Comment by glasser** [comment]
This is a published library (`@apollo/gateway`) that anyone can use, and we are planning to add OT spans to it. So you suggest a direct dependency and having our code call `trace.getTracer`?

**Comment by dyladan** [comment]
> What do you mean by "even though you can't use the SDK with it right now" though?

The currently released SDK is pinned to the 0.x API. A new SDK/contrib release is required for use with the 1.0 API (this week).

> This is a published library (`@apollo/gateway`) that anyone can use, and we are planning to add OT spans to it. So you suggest a direct dependency and having our code call `trace.getTracer('default')`?

As the library author this is really up to you. We decided to go with peer dependency in the instrumentations and sdk to allow the users to explicitly pick a version of the API if they want to. For us this makes sense because there is no reason to install these components if you're not going to install the API. For libraries which depend on the API it is a little harder though because you probably don't want to make your users install the API manually if they're not otel users. Depending on the API directly should work fine, but just makes dependency management slightly more automatic and harder to control.

**Comment by dyladan** [comment]
>This is a published library (@apollo/gateway) that anyone can use, and we are planning to add OT spans to it. So you suggest a direct dependency and having our code call trace.getTracer('default')?

Shorter answer is "yes"

**Comment by vmarchaud** [comment]
> I'm confused about what a library like this should be passing as the name to getTracer

You should use your library name and give version also, this is attached to every span created within your lib (you have more info there: https://github.com/open-telemetry/opentelemetry-specification/pull/616)
